### ISSUE TITLE
ARDUINO_ARCH_HC32 defined to late?

### ISSUE THREAD

**Issue by ellensp** [shadow578/framework-arduino-hc32f46x#17]
Please bare with me my terminology is probably all wrong (frameworks are above my pay grade!) 

With Marlin when you build (for any system) one of the stages it goes through is

`.platformio/packages/toolchain-atmelavr/bin/avr-g++" -D__MARLIN_FIRMWARE__ -DNDEBUG -D__MARLIN_DEPS__ -w -dM -E -x c++ buildroot/share/PlatformIO/scripts/common-dependencies.h
`
The generates a list of all used defines for use in buildroot/share/PlatformIO/scripts/preprocessor.py  

On all other platforms this 'builds' without issue
On Framework-arduino-hc32f46x this triggers the error in Marlin/src/pins/hc32f4/env_validate.h

```
#ifndef ARDUINO_ARCH_HC32
  #error "Oops! Select an HC32F460 board in 'Tools > Board.'"
#endif
``` 

This breaks buildroot/share/PlatformIO/scripts/preflight-checks.py 
Normally if you set #define MOTHERBOARD to a board that requires a different ENV: this will tell you the valid envronments you can use.

A worknig eg:   start with stock configs and set #define MOTHERBOARD BOARD_BTT_SKR_V1_4
click build and marlin tells you 
"Error: Build environment 'mega2560' is incompatible with BOARD_BTT_SKR_V1_4. Use one of these environments: LPC1768"

With a HC32 board: 
eg #define MOTHERBOARD BOARD_AQUILA_V101
results in   #error "Oops! Select an HC32F460 board in 'Tools > Board.'"

removing the check from Marlin/src/pins/hc32f4/env_validate.h gives the expected results 
"Error: Build environment 'mega2560' is incompatible with BOARD_AQUILA_V101. Use one of these environments: HC32F460C_aquila_101"
But this is not a solution. 

Could you look into this?

Thanks.

**Comment by shadow578** [comment]
if i've understood this issue correctly, the issue is that marlins preflight check fails, correct?

if so, this would probably not be a issue in the arduino framework.
from what i've seen in other `env_validate.h` files (e.g. [`lpc1769/env_validate.h`](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/pins/lpc1769/env_validate.h), they use `NOT_TARGET` instead of directly checking if the macro is defined.


the `NOT_TARGET` macro is itself defined as:
```cpp
#ifdef __MARLIN_DEPS__
  #define NOT_TARGET(V...) 0
#else
  #define NOT_TARGET NONE
#endif
```
(see [`pins.h`](https://github.com/MarlinFirmware/Marlin/blob/d0d229e380d9b5c82dfeef4980f5b7b4745e933d/Marlin/src/pins/pins.h#L120))


> But this is not a solution.

basically, it is. 
changing the check in `hc32f4/env_validate.h` to the following fixes the behaviour:
```cpp
#if NOT_TARGET(ARDUINO_ARCH_HC32)
  #error "Oops! Select an HC32F460 board in 'Tools > Board.'"
#endif
```


i'll create a PR in the marlin repo to add this change, referencing this issue.

---

as a sidenote, this issue seems to be caused because in the `mega2560` environment (or any non-hc32 environment), the hc32 arduino core isn't invoked at all. 
so no way to fix this in the core...

**Comment by ellensp** [comment]
I see how it woks now, I was so close and yet miles off!
Thanks for fixing this.
